File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} list1
 * @param {ListNode} list2
 * @return {ListNode}
 */
var mergeTwoLists = function(list1, list2) {
   
    let dummyList = new ListNode(-10000);
    let tempNode =dummyList; //point the dummyList
    while(list1 && list2){
        if(list1.val<list2.val){
            tempNode.next=list1;
            list1=list1.next;
            console.log("tempNode inside while",tempNode);
            console.log("dummy after while",dummyList);
        }else{
            tempNode.next=list2;
            list2=list2.next;  
            console.log("tempNode inside while",tempNode);
            console.log("dummy inside while",dummyList);          
        }
        tempNode=tempNode.next;  
            console.log("tempNode after while",tempNode);
            console.log("dummy after while",dummyList);  
        
    }
    if(list1 ==null) tempNode.next=list2;
    if(list2==null) tempNode.next =list1;
    console.log("tempNode",tempNode);
    console.log("dummy",dummyList);
    return dummyList.next;
//    let headNode ;
//    let tempNode;
//    if(!list1 ){
//        return list2;
//    }else if(!list2){
//        return list1;
//   }
//    if(list1.val < list2.val){
//        headNode= list1;
//        tempNode = headNode;
//        list1=list1.next;
//    }else{
//        headNode= list2;
//        tempNode = headNode;
//        list2=list2.next;
//    }
//    while(list1 && list2){
//        if(list1.val < list2.val){
//            tempNode.next = list1;
//            list1=list1.next;
//        }else{
//            tempNode.next = list2;
//            list2=list2.next;
//        }
//       // ba7arek tempNode evey loop
//        tempNode= tempNode.next;
//    }
//    if(list1 ==null) tempNode.next=list2;
//    if(list2==null) tempNode.next =list1;
//  return headNode; 
//
//    
};
//
//
//Time Complexity: O(N + M), where N and M are the size of list1 and list2 respectively
//Space Complexity: O(1)

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        #list1 = [1,2,4], list2 = [1,3,4]
        #         |                |         temp.next=1     list1=1    list2=1
        #         |                  |       temp.next=1     list1=1    list2=3
        #           |                |       temp.next=2     list1=2    list2=3
        #             |              |       temp.next=3     list1=4    list2=3
        #             |                |     temp.next=4     list2=4    list2=4
        #             |                   |  temp.next=4     list2=4    list2=None

        #edge case
        if not list1:
            return list2
        elif not list2:
            return list1    
            
        temp=ListNode()

        if list1.val<list2.val:
            head=list1
        else:
            head=list2
                 
        while list1 and list2:
            if list1.val<list2.val:
                temp.next=list1
                temp=temp.next
                list1=list1.next
            else:
                temp.next=list2
                temp=temp.next
                list2=list2.next
        if list2:
            temp.next=list2
        if list1:
            temp.next=list1            
        return head            

               

        